src/test/java/apoc/concurrent/SemaphoreProceduresTest.java

```diff
@@ -0,0 +1,125 @@
+    }
+
+    @Test
+    public void testThatSemaphoreSerializesOperations() throws InterruptedException {
+        long duration = 200;
+        String cypher = "call apoc.concurrent.runInSemaphore($semaphoreName, 'call apoc.util.sleep($duration) return 1', {duration: $duration})";
+        Map<String, Object> params = MapUtil.map(
+                "duration", duration,
+                "semaphoreName", "single"
+        );
+        db.executeTransactionally(cypher, params, Iterators::single); // ensure query plan cache gets populated
+        long now = System.currentTimeMillis();
+
+        Set<Thread> threads = Iterators.asSet(
+                new Thread(() -> db.executeTransactionally(cypher, params, Iterators::single)),
+                new Thread(() -> db.executeTransactionally(cypher, params, Iterators::single))
+        );
+
+        threads.forEach(Thread::start);
+        threads.forEach(thread -> {
+            try {
+                thread.join();
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        });
+
+        assertTrue( (System.currentTimeMillis()- now) >= 2*duration);
+        long availablePermits = TestUtil.singleResultFirstColumn(db, "return apoc.concurrent.availablePermits($semaphoreName)", params);
+        assertEquals(1l, availablePermits);
+    }
+
+    @Test
+    public void testSemaphoreReleasedOnFailure() throws InterruptedException {
+        String cypher = "call apoc.concurrent.runInSemaphore($semaphoreName, 'call does.not.exist() return 1')";
+        Map<String, Object> params = MapUtil.map("semaphoreName", "single");
+        Set<Thread> threads = Iterators.asSet(
+                new Thread(() -> db.executeTransactionally(cypher, params, Iterators::single)),
+                new Thread(() -> db.executeTransactionally(cypher, params, Iterators::single))
+        );
+
+        threads.forEach(Thread::start);
+        threads.forEach(thread -> {
+            try {
+                thread.join();
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        });
+
+        long availablePermits = TestUtil.singleResultFirstColumn(db, "return apoc.concurrent.availablePermits($semaphoreName)", params);
+        assertEquals("Ensure 'release' has been called.", 1l, availablePermits);
+    }
+
+    @Test(timeout = 10000)
+    public void testRunInSemaphoreIsKillable() throws InterruptedException {
+        Thread thread = new Thread(() -> {
+            db.executeTransactionally("call apoc.concurrent.runInSemaphore($semaphoreName, 'return 1')", Collections.singletonMap("semaphoreName", SEMAPHORE_ZERO), Iterators::single);
+        });
+        thread.start();
+        Thread.sleep(100);  // await planning
+
+        // kill all active queries
+        DependencyResolver dependencyResolver = ((GraphDatabaseAPI) db).getDependencyResolver();
```

Comment:
should we test this also from a users perspective? like listQueries/killQuery?

what happens if the inner query is killed?